### PR TITLE
Add rekey multisig support for algokey

### PR DESCRIPTION
**Pull request by algorotem**
Currently, it's impossible to sign a transaction for an account that was rekeyed to a multisig account. This PR adds the ability to convert a regular transaction that goal creates to a one that contains all the information needed for signing using `algokey multisig`.

### CODE REVIEW

**Comment by ryanRfox** [comment]
I successfully signed and sent a transaction from an account which was --rekey-to multisig. Below $ACCOUNT_A has its _auth-addr_ set to $BC_T1 which is a multisig of $ACCOUNT_B $ACCOUNT_C threshold 1. The unsigned transaction sends from $ACCOUNT_A which requires either $ACCOUNT_B or $ACCOUNT_C to sign. The `algokey multisig convert` command adds the "msig" so the "s" can be added by `algokey multisig`. Not the best UX, but my account is not locked :) 

# RESET: rm ${TEMPDIR}/test5*
goal account list -d $DATA
${GOAL_CMD} clerk send --from ${ACCOUNT_A} --to ${ACCOUNT_B} --amount 100000 --out ${TEMPDIR}/test5.utxn
algokey multisig convert --txfile ${TEMPDIR}/test5.utxn --params "1 $ACCOUNT_B $ACCOUNT_C"
#FAIL: goal clerk inspect ${TEMPDIR}/test5.utxn
cat $TEMPDIR/test5.utxn | msgpacktool -d -b32
algokey multisig --txfile ${TEMPDIR}/test5.utxn --outfile ${TEMPDIR}/test5.stxn --keyfile ${TEMPDIR}/account_b
#FAIL: goal clerk inspect ${TEMPDIR}/test5.stxn
cat $TEMPDIR/test5.stxn | msgpacktool -d -b32
${GOAL_CMD} clerk rawsend --filename ${TEMPDIR}/test5.stxn

Notice: `goal clerk inspect` does not like the converted transaction files; use `msgpacktool` to view them.

File: cmd/algokey/multisig.go [hunk 1]
```diff
@@ -93,3 +105,78 @@ var multisigCmd = &cobra.Command{
 		}
 	},
 }
+
+var convertCmd = &cobra.Command{
+	Use:   "convert -t [transaction file] -p \"[threshold] [PK1] [PK2] ...\"",
+	Short: "Adds the necessary fields to a transaction that is sent from an account to was rekeied to a multisig account.",
```
Comment (tsachiherman) [inline cmd/algokey/multisig.go:111]: Could you drop the "." from the end of the line ? We don't add "." on other commands.

File: cmd/algokey/multisig.go [hunk 2]
```diff
@@ -93,3 +105,78 @@ var multisigCmd = &cobra.Command{
...
+		if basics.Address(addr) == stxn.Txn.Sender {
+			fmt.Fprintf(os.Stderr, "The sender at the msig address should not be the same: %v\n", err)
+			os.Exit(1)
+		}
+		stxn.AuthAddr = basics.Address(addr)
+
+		// Write the txn
+		outBytes = append(outBytes, protocol.Encode(&stxn)...)
+
+		err = ioutil.WriteFile(multisigTxfile, outBytes, 0600)
```
Comment (tsachiherman) [inline cmd/algokey/multisig.go:176]: maybe use `writeFile` so that we can support `stdoutFilenameValue` ?

File: cmd/algokey/multisig.go [hunk 3]
```diff
@@ -93,3 +105,78 @@ var multisigCmd = &cobra.Command{
...
 }
+
+var convertCmd = &cobra.Command{
+	Use:   "convert -t [transaction file] -p \"[threshold] [PK1] [PK2] ...\"",
+	Short: "Adds the necessary fields to a transaction that is sent from an account to was rekeied to a multisig account.",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, _ []string) {
+
+		// Read Transaction
+		txdata, err := ioutil.ReadFile(multisigTxfile)
```
Comment (tsachiherman) [inline cmd/algokey/multisig.go:116]: does it make sense to use `readFile` here ?

File: cmd/algokey/multisig.go (hunk 1 above)
Comment (tsachiherman) [inline cmd/algokey/multisig.go:111]: typo : `rekeied` -> `rekeyed`
           `account to was` -> `account that was`

File: cmd/algokey/multisig.go (hunk 2 above)
Comment (algorotem) [reply cmd/algokey/multisig.go:176]: I'm not sure I follow. It does use `WriteFile`. Do you mean to let the user choose the name?

**Comment by brianolson** [comment]
Why is this being made a part of `algokey` instead of `goal` ? `goal` is where we do the rest of the txn operations.

**Comment by algorotem** [comment]
> Why is this being made a part of `algokey` instead of `goal` ? `goal` is where we do the rest of the txn operations.

in addition, just happened to be first

File: cmd/algokey/multisig.go [hunk 4]
```diff
@@ -109,8 +109,8 @@ var multisigCmd = &cobra.Command{
 	},
 }
 
-var convertCmd = &cobra.Command{
-	Use:   "convert -t [transaction file] -p \"[threshold] [PK1] [PK2] ...\"",
```
Comment (ryanRfox) [inline cmd/algokey/multisig.go]: I'd like to see the help be more explicit on the address ordering:

	Use:   "append-auth-addr -t [transaction file] -p \"[threshold] [Address 1] [Address 2] ... [Address n]\"",

File: cmd/algokey/multisig.go [hunk 5]
```diff
@@ -50,11 +50,11 @@ func init() {
 	multisigCmd.Flags().StringVarP(&multisigOutfile, "outfile", "o", "", "Transaction output filename")
 	multisigCmd.MarkFlagRequired("outfile")
 
-	convertCmd.Flags().StringVarP(&msigParams, "params", "p", "", "Multisig params -  Threshold PK1 PK2 ...")
-	convertCmd.MarkFlagRequired("params")
-	convertCmd.Flags().StringVarP(&multisigTxfile, "txfile", "t", "", "Transaction input filename")
-	convertCmd.MarkFlagRequired("txfile")
-	convertCmd.Flags().StringVarP(&multisigOutfile, "outfile", "o", "", "Transaction output filename. If not specified, the original file will be modified")
+	appendAuthAddrCmd.Flags().StringVarP(&msigParams, "params", "p", "", "Multisig pre image parameters - [threshold] [Address 1] [Address 2]")
```
Comment (ryanRfox) [inline cmd/algokey/multisig.go:53]: I'd like to see the help be more explicit on the address ordering:

	appendAuthAddrCmd.Flags().StringVarP(&msigParams, "params", "p", "", "Multisig pre image parameters - [threshold] [Address 1] [Address 2] .. [Address n]")

Comment (tsachiherman) [reply cmd/algokey/multisig.go:53]: @ryanRfox,

We need to converge the way we present multiple optional parameters. For other UXs, we've used the following:
"inspect [input file 1] [input file 2]..."
"new [address 1] [address 2]..."
"merge -o [merged transaction file] [input file 1] [input file 2]..."

I don't mind picking any particular "pattern", but would like to ensure they are all the same.

Comment (ryanRfox) [reply cmd/algokey/multisig.go:53]: I agree, we must use a consistent pattern. In this case, please use what is established elsewhere and disregard my comment on this PR.

**Pull request merged**